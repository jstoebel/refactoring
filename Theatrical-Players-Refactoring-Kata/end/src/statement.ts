import {IInvoice, IPlays} from './types'

export default function statement (invoice: IInvoice, plays: IPlays) { // long function
    let totalAmount = 0; // mutable
    let volumeCredits = 0; // mutable
    let result = `Statement for ${invoice.customer}\n`; // mutable, myterious name
    const currencyFormatter = new Intl.NumberFormat("en-US",
        { style: "currency", currency: "USD",
            minimumFractionDigits: 2 }).format; // myterious name

    const performancesData = invoice.performances.map((perf) => {
        const play = plays[perf.playID];
        let thisAmount = 0; // mutable, mysterious. the cost for a single performance
        switch (play.type) { // switch!
            case "tragedy":
                thisAmount = 40000;
                if (perf.audience > 30) {
                    thisAmount += 1000 * (perf.audience - 30);
                }
                break;
            case "comedy":
                thisAmount = 30000;
                if (perf.audience > 20) {
                    thisAmount += 10000 + 500 * (perf.audience - 20);
                }
                thisAmount += 300 * perf.audience;
                break;
            default:
                throw new Error(`unknown type: ${play.type}`);
        }
        // add volume credits
        volumeCredits += Math.max(perf.audience - 30, 0);
        // add extra credit for every ten comedy attendees
        if ("comedy" === play.type) volumeCredits += Math.floor(perf.audience / 5); // random business logic
        // print line for this order

        return {
            text: `${play.name}: ${currencyFormatter(thisAmount/100)} (${perf.audience} seats)`,
            performanceCost: thisAmount
        }
    })

    totalAmount = performancesData.map((perf) => perf.performanceCost).reduce((prev, curr) => prev + curr, 0)

    result += performancesData.map((perf) => ` ${perf.text}`).join('\n')
    result += '\n'
    result += `Amount owed is ${currencyFormatter(totalAmount/100)}\n`;
    result += `You earned ${volumeCredits} credits\n`;
    return result;
}
