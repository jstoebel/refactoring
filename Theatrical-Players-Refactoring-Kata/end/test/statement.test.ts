import statement from '../src/statement';
import fs from 'fs'

test('example statement', () => {
    const invoice = JSON.parse(fs.readFileSync('test/invoice.json', 'utf8'));
    const plays = JSON.parse(fs.readFileSync('test/plays.json', 'utf8'));
    expect(statement(invoice, plays)).toMatchSnapshot();
});

test('statement with new play types', () => {
    const invoice = JSON.parse(fs.readFileSync('test/invoice_new_plays.json', 'utf8'));
    const plays = JSON.parse(fs.readFileSync('test/new_plays.json', 'utf8'));

    // TODO: change to snap shot when it passes the first time
    const newPlaysExpected = fs.readFileSync('test/new_plays_output.txt', 'utf8')
    // expect(statement(invoice, plays)).toEqual(newPlaysExpected)
    expect(statement(invoice, plays)).toMatchSnapshot();
});

test('html output', () => {
    const invoice = JSON.parse(fs.readFileSync('test/invoice.json', 'utf8'));
    const plays = JSON.parse(fs.readFileSync('test/plays.json', 'utf8'));
    expect(statement(invoice, plays, {output: 'html'})).toMatchSnapshot()
})
