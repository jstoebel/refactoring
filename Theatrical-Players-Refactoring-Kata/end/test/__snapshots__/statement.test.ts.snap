// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example statement 1`] = `
"Statement for BigCo
 Hamlet: $650.00 (55 seats)
 As You Like It: $580.00 (35 seats)
 Othello: $500.00 (40 seats)
Amount owed is $1,730.00
You earned 47 credits
"
`;

exports[`html output 1`] = `
"<div>
  <h1>Statement for BigCo</h1>
  <ol>
      <li>Hamlet: $650.00 (55 seats)</li>
      <li>As You Like It: $580.00 (35 seats)</li>
      <li>Othello: $500.00 (40 seats)</li>
  </ol>
  <div>Amount owed is $1,730.00</div>
  <div>You earned 47 credits</div>
</div>
"
`;

exports[`statement with new play types 1`] = `
"Statement for BigCoII
 Henry V: $676.00 (53 seats)
 As You Like It: $740.00 (55 seats)
Amount owed is $1,416.00
You earned 55 credits
"
`;
